Unified attention CK Tile kernel

juuso-oskari opened the pull request:
Authors: @Chi-Chu319 @juuso-oskari 

This PR implements a unified attention kernel written in CK Tile. It builds on top of the fmha_v3 (composable_kernel/example/ck_tile/01_fmha) with the pipeline largely remaining the same. This PR implements the following features introduced in [Triton unified attention kernel](https://github.com/ROCm/aiter/blob/355_wip/aiter/ops/triton/unified_attention.py):

reduced launch grid size at composable_kernel/example/ck_tile/01_unified_attention/unified_attention_impl.hpp
```
// args.num_tokens is the cumulative amount of tokens from all sequences
index_t total_num_q_blocks = args.num_tokens / BLOCK_Q + args.num_seqs;
dim3 grids            = Kernel::GridSize2D(args.num_kv_heads, total_num_q_blocks);
return launch_kernel(config, make_kernel<kBlockPerCu>(Kernel{}, grids, blocks, 0, kargs));
```
This is significantly less amount of programs launched compared to before grid=(num_seqs, max_seqlen // BLOCK_M, num_q_heads), which contained lots of empty programs (not all sequences are of length max_seqlen).

But since now the current sequence index cannot be taken from the program id, we need to do a binary search at the beginning of the kernel to find our sequence index (used to index sequence length; needed for determining innerloop length).

This is implemented at composable_kernel/include/ck_tile/ops/unified_attention/kernel/unified_attention_kernel.hpp:
```
// Binary search to find the sequence index for a given global index
CK_TILE_DEVICE static constexpr ck_tile::index_t
find_seq_idx(const int32_t* query_start_len_ptr,
                ck_tile::index_t target_idx,
                ck_tile::index_t num_seqs,
                ck_tile::index_t block_q,
                bool use_q_block_mode)
{
    ck_tile::index_t left = 0;
    ck_tile::index_t right = num_seqs;
    while (left < right)
    {
        ck_tile::index_t mid = (left + right) / 2;
        ck_tile::index_t val = query_start_len_ptr[mid];
        ck_tile::index_t mid_val = use_q_block_mode ? (val / block_q + mid) : val;
        
        if (mid_val <= target_idx)
        {
            left = mid + 1;
        }
        else
        {
            right = mid;
        }
    }
    return left - 1;
}
// usage inside the kernel
const auto [kv_head_idx, q_block_global_idx] = GetTileIndex(pid, kargs);
// grid size is (num_kv_heads, total_num_q_blocks)
// total_num_q_blocks = q.shape[0] // BLOCK_Q + num_seqs
// q.shape[0] is total number of query tokens across all batches
const index_t seq_idx = find_seq_idx(
    kargs.query_start_len_ptr, q_block_global_idx, kargs.num_seqs, BLOCK_Q, true
); // which seq am I
```
In order to process more query tokens per load in decode settings (where sequence length is small, often only 1), we group query tokens in the head dim. Up to num_queries_per_kv query tokens share the same key/value token (CQA-setting). The total number of grouped tokens for a tile load is BLOCK_M = BLOCK_Q * num_queries_per_kv.

We do this in the kernel implementation by transforming the tensor view for Q in dram:
```
const auto q_dram = [&]() {
    const auto q_dram_base = make_naive_tensor_view<address_space_enum::global>(
        q_ptr,
        make_tuple(cur_batch_query_len, num_queries_per_kv, HEAD_SIZE),
        make_tuple(kargs.query_stride_0, kargs.query_stride_1, 1),
        number<UnifiedAttentionPipeline::kAlignmentQ>{},
        number<1>{});

    const auto q_dram_pad = pad_tensor_view( // aling seqlen with BLOCK_Q and head dim with HEAD_SIZE_PADDED
        q_dram_base,
        // block sizes
        make_tuple(BLOCK_Q, 1, HEAD_SIZE_PADDED),
        sequence<true, false, kPadHeadDimQ>{}
    ); // pads to (seq_len_padded, num_head_q, HEAD_SIZE_PADDED)

    const auto q_dram_merged = transform_tensor_view(
                q_dram_pad,
                make_tuple(
                    make_merge_transform(
                        make_tuple(query_len_padded, num_queries_per_kv)
                    ),
                    make_pass_through_transform(HEAD_SIZE_PADDED)
                ),
                make_tuple(sequence<0, 1>{}, sequence<2>{}),
                make_tuple(sequence<0>{}, sequence<1>{})
    ); // flattens the first two dims, head idx is the fastest changing dim in the merged dim
    return q_dram_merged;
}();
```
This way, pipeline can remain untouched and use the BLOCK_M as its tile size.

## build

```
cd ~/composable_kernel
# in the root
mkdir build && cd build
# you can replace "gfx950" with the appropriate architecture (for example gfx90a or gfx942) or leave it blank
../script/cmake-ck-dev.sh .. "gfx950" -G Ninja
ninja tile_example_unified_attention -j64
./bin/tile_example_unified_attention
```

## Benchmark comparison against FAv3 and Triton Unified Attention

**Prefill batch: b=1, h=64, d=128, seqlen_q/k=16384, causal=1**

Ours: 548 TFLOPS
```
# need to manually change the num_queries_per_kv from the example files to 1
jukorhon@asrock-1w300-e0-3:~/composable_kernel/build$ ./bin/tile_example_unified_attention -h_k=64 -query_lens=16384 -kv_lens=16384 -varlen=0
[bf16|] b:1, h:64/64, d:128, scale_s:0.0883883, query_lens:[16384], kv_lens:[16384], mask:causal mask, 8.01874065 ms, 548.50 TFlops, 0.07 TB/s
```
fmha_pagedkv: 531 TFLOPS
```
jukorhon@asrock-1w300-e0-3:~/composable_kernel/build$ ./bin/tile_example_fmha_fwd -h=64 -b=1 -mask='b' -s=16384 -v=0 -prec=bf16 -page_block_size=128
[bf16|batch|bhsd] b:1, h:64/64, s:16384/16384, d:128/128, scale_s:0.0883883, bias:n, p_drop:0, lse:0, qscale:n, mask:b(-1:0), v:r, page_block_size:128, 8.273 ms, 531.63 TFlops, 129.78 GB/s
```
fmha_v3: 654 TFLOPS
```
jukorhon@asrock-1w300-e0-3:~/composable_kernel/build$ ./bin/tile_example_fmha_fwd_v3 -h=64 -q_eff_lens=16384 -kv_eff_lens=16384 -b=1 -causal=1 -s=16384 -v=0
[fp16|bshd] b:1, h:64/64, s:16384/16384, d:128, scale_s:0.0883883, mask:b(-1:0), 6.719 ms, 654.66 TFlops
```
FAv3 Triton [blogpost](https://amd.atlassian.net/wiki/spaces/MLSE/pages/1048754871/Episode+1+FAv3+on+MI350+Unleashed): 654 TFLOPS

Unified Attention Triton (at branch [bench-unified-attention](https://github.com/ROCm/aiter/tree/bench-unified-attention):
```
(py_3.10) tianxiwu@asrock-1w300-e0-3:~/rocm/aiter$ python op_tests/op_benchmarks/triton/bench_unified_attention.py  -hq 64 -hk 64 -b 1 -sq 16384 -sk 16384 -d 128 -unified_attention -causal true --layout "thd"
[aiter] import [module_aiter_enum] under /home/tianxiwu/rocm/aiter/aiter/jit/module_aiter_enum.so
bench_unified_attention:
   BATCH    HQ    HK  N_CTX_Q  N_CTX_K  D_HEAD  D_HEAD_V  fwd(TFLOPS)
0    1.0  64.0  64.0  16384.0  16384.0   128.0     128.0   472.475302
```

**Mixed batch: b=8, h_q / h_k=128/8, d=128, seqlen_q=1024,1,32,256,1,1,4096,4096, seqlen_k=16384, causal=1**

Ours: 674 TFLOPS
```
# need to manually change the num_queries_per_kv from the example files to 16
jukorhon@asrock-1w300-e0-3:~/composable_kernel/build$ ./bin/tile_example_unified_attention -h_k=8 -query_lens=1024,1,32,256,1,1,4096,4096 -kv_lens=16384,16384,16384,16384,16384,16384,16384,16384 -varlen=0
[0/2] Re-checking globbed directories...
[7/8] Linking CXX executable bin/tile_example_unified_attention
[bf16|] b:8, h:128/8, d:128, scale_s:0.0883883, query_lens:[1024,1,32,256,1,1,4096,4096], kv_lens:[16384,16384,16384,16384,16384,16384,16384,16384], mask:causal mask, 13.45901966 ms, 674.07 TFlops, 0.05 TB/s
```
Unified Attention Triton (at branch [bench-unified-attention](https://github.com/ROCm/aiter/tree/bench-unified-attention)): 562 TFLOPS
```
(py_3.10) jukorhon@asrock-1w300-e0-3:~/aiter$ python op_tests/op_benchmarks/triton/bench_unified_attention.py -causal true -d 128 -hq 32 -hk 8 -sq 128,1,1,1024,1,1,4096,4096 -sk 16384 -unified_attention --layout "thd"
[aiter] import [module_aiter_enum] under /home/jukorhon/aiter/aiter/jit/module_aiter_enum.so
bench_unified_attention:
   BATCH  HQ  HK                     N_CTX_Q  N_CTX_K  D_HEAD  D_HEAD_V  fwd(TFLOPS) (TFLOPS)
0      8  32   8  128,1,1,1024,1,1,4096,4096    16384     128       128            562.924279
```

Chi-Chu319 committed:
intial commit

Chi-Chu319 committed:
unified attention rename

juuso-oskari committed:
transform q tensor view

Chi-Chu319 committed:
refactor

Chi-Chu319 committed:
refactor. and fixed q transformation

Chi-Chu319 committed:
Some refactor

Chi-Chu319 committed:
refactor

juuso-oskari committed:
refactor the q tensor view transformation

juuso-oskari committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

juuso-oskari committed:
refactor the q tensor view transformation

Chi-Chu319 committed:
kv tensor view

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
stride fix

juuso-oskari committed:
fix

juuso-oskari committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

juuso-oskari committed:
add commenting

Chi-Chu319 committed:
o ptr and window

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
kv tensor view and initial window

juuso-oskari committed:
fix q window origin

juuso-oskari committed:
fix q window

Chi-Chu319 committed:
pipeline api

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
o window change

Chi-Chu319 committed:
kv paging

Chi-Chu319 committed:
Comments

juuso-oskari committed:
fix seq_len -> cur_batch_query_len

juuso-oskari committed:
merge

Chi-Chu319 committed:
Example boostrap

juuso-oskari committed:
correct masking by transforming y_idx = y_idx / num_queries_per_kv

juuso-oskari committed:
merge

juuso-oskari committed:
use correct mask in kernel

juuso-oskari committed:
fix mask

juuso-oskari committed:
comment

juuso-oskari committed:
fix order in mask caller

Chi-Chu319 committed:
example

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
fixing args

juuso-oskari committed:
modified cmake files at unified attention example. Now cmake works, but getting compile errors (expected atm)

Chi-Chu319 committed:
Compiling fix

juuso-oskari committed:
fixing compile errors...

juuso-oskari committed:
fixing compile errors...

Chi-Chu319 committed:
More compilation fixes

juuso-oskari committed:
change to BLOCK_M in shape definitions

juuso-oskari committed:
fixing bugs

Chi-Chu319 committed:
fixed example

Chi-Chu319 committed:
block table stride fix

juuso-oskari committed:
fixing bugs

Chi-Chu319 committed:
fix the vector max

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
removed redundent code

Chi-Chu319 committed:
Fixed pipeline args

Chi-Chu319 committed:
const blockq

Chi-Chu319 committed:
Fixed block Q with M

Chi-Chu319 committed:
Fixed block Q with M

Chi-Chu319 committed:
Debugging window size

Chi-Chu319 committed:
fixed window creation number<>{}

Chi-Chu319 committed:
Added block q

Chi-Chu319 committed:
Merge branch 'develop' into tianxing/unified-attention

Chi-Chu319 committed:
revert change on fmha

Chi-Chu319 committed:
More compilation fix

Chi-Chu319 committed:
cmake list update

Chi-Chu319 committed:
fixed block sizes

Chi-Chu319 committed:
block_shape fixes

Chi-Chu319 committed:
block_shape fixes

Chi-Chu319 committed:
Fixed impl

juuso-oskari committed:
fix to example

juuso-oskari committed:
add handling for -1 k heads arg

juuso-oskari committed:
refactor to clearer BLOCK Q logic

juuso-oskari committed:
deving the test...

Chi-Chu319 committed:
remove unneeded args

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
Assert block_size num_queries_per_kv

Chi-Chu319 committed:
flops and mem calculation

Chi-Chu319 committed:
mem calculation fixed

juuso-oskari committed:
fixing

Chi-Chu319 committed:
Various fixes

Chi-Chu319 committed:
ref data copying

juuso-oskari committed:
fix

Chi-Chu319 committed:
Pipeline minor fixes

Chi-Chu319 committed:
Fixed the block_table

Chi-Chu319 committed:
correct results

Chi-Chu319 committed:
causal mask fix

Chi-Chu319 committed:
non zero calculation fix

Chi-Chu319 committed:
fmha v3 flops calculation

Chi-Chu319 committed:
simplified kernel pid logic

Chi-Chu319 committed:
refined benchmarking

Chi-Chu319 committed:
Merge branch 'tianxing/unified-attention' of https://github.com/ROCm/composable_kernel into tianxing/unified-attention

Chi-Chu319 committed:
Merge branch 'develop' into tianxing/unified-attention

Chi-Chu319 committed:
Merge branch 'develop' into tianxing/unified-attention

juuso-oskari committed:
adding PAGE_BLOCK_SIZE >= BLOCK_SIZE optionality, now it regresses perf when it should improve?

juuso-oskari committed:
fix

samremes commented on include/ck_tile/ops/unified_attention/block/block_attention_bias_enum.hpp:
Did you make any modifications to this, or is this all same as `composable_kernel/include/ck_tile/ops/fmha/block/block_attention_bias_enum.hpp`? Let's try to avoid code duplication as much as reasonably possible, you can include files from other ops too, if needed.

samremes commented on include/ck_tile/ops/unified_attention/block/block_dropout.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/block/block_masking.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/block/block_position_encoding.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/block/block_rotary_embedding.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/block/page_block_navigator.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/block/variants.hpp:
Check if you can include this also from `composable_kernel/include/ck_tile/ops/fmha/block/`.

samremes commented on include/ck_tile/ops/unified_attention/kernel/unified_attention_kernel.hpp:
Maybe this kernel could be placed under ops/fmha? Since they share a lot of functionality under `block/` directory too. New pipelines under `pipeline` could be added to `ops/fmha/pipeline`.

Chi-Chu319 replied:
I would probably keep them seperate still. There are also some features going to be added to the unified attention eventually they are going to deviate

samremes commented on include/ck_tile/ops/unified_attention/kernel/unified_attention_kernel.hpp:
```diff
@@ -0,0 +1,456 @@
+    static constexpr bool kPadHeadDimQ = UnifiedAttentionPipeline::kPadHeadDimQ;
+    static constexpr bool kPadHeadDimV = UnifiedAttentionPipeline::kPadHeadDimV;
+
+    // TODO add yjese
```
yjese?

samremes commented on include/ck_tile/ops/unified_attention/kernel/unified_attention_kernel.hpp:
```diff
@@ -0,0 +1,456 @@
+
+        assert(BLOCK_M / num_queries_per_kv == BLOCK_Q);
+
+        // const index_t BLOCK_Q = BLOCK_M / num_queries_per_kv;
```
Remove commented code.

samremes commented on include/ck_tile/ops/unified_attention/pipeline/unified_attention_pipeline_enum.hpp:
This could be also just included from fmha.

samremes commented on include/ck_tile/ops/unified_attention/pipeline/tile_unified_attention_shape.hpp:
```diff
@@ -0,0 +1,72 @@
+    static constexpr index_t BLOCK_SIZE = BlockTile::at(number<2>{}); //  BLOCK size for K seqlen
+    static constexpr index_t HEAD_SIZE  = BlockTile::at(number<3>{}); //  BLOCK size for K seqlen
+
+    // static constexpr index_t kQKHeaddim =
```
remove commented code

samremes commented on include/ck_tile/ops/unified_attention/pipeline/unified_attention_pipeline_default_policy.hpp:
```diff
@@ -0,0 +1,605 @@
+            {
+                /// NOTICE: in order to use load_tile_transpose() later for V tile, we cannot use
+                /// WarpGemmMfmaBf16Bf16F32M32N32K16SwizzleBTransposedCDistribution here
+                return WarpGemmMfmaBf16Bf16F32M32N32K16TransposedCDistribution<>{};
```
I wonder if it would be somehow possible to use the WarpGemmDispatcher also here, could make it easier to extend for FP8 etc. 

Would probably need some changes to the dispatcher, like adding some extra parameters to accommodate more fine-grained selection for special cases like this. But I guess many fmha pipelines do it like this, or at least the v3 does it like this.

samremes commented on include/ck_tile/ops/unified_attention/kernel/unified_attention_kernel.hpp:
```diff
@@ -0,0 +1,456 @@
+    // BLOCK_Q = BLOCK_M // num_queries_per_kv
+    // BLOCK_Q is the block size for q seqlen
+    /// static constexpr index_t BLOCK_Q = UnifiedAttentionPipeline::BLOCK_Q;
+    static constexpr index_t BLOCK_M = UnifiedAttentionPipeline::BLOCK_M;
```
Could we follow the more common naming convention, e.g. `kBlockM` here for these?

samremes commented on include/ck_tile/ops/unified_attention/pipeline/unified_attention_pipeline.hpp:
```diff
@@ -0,0 +1,1244 @@
+struct CoreLoopScheduler;
+
+template <typename PipelineProblem>
+struct CoreLoopScheduler<PipelineProblem, /*kIsMasking=*/true>
```
It could be probably a good idea to include these from the v3 pipeline. They maybe getting changed there, as the v3 pipeline is still somehow in progress, not fully finalized.

samremes commented on include/ck_tile/ops/unified_attention/pipeline/unified_attention_pipeline.hpp:
```diff
@@ -0,0 +1,1244 @@
+                 : [lhs] "v"(lhs), [rhs] "v"(rhs));
+    return result;
+}
+} // namespace detail
```
Also all of these could be put into some common place for v3 and unified pipelines.

samremes commented on include/ck_tile/ops/unified_attention/pipeline/unified_attention_pipeline.hpp:
```diff
@@ -0,0 +1,1244 @@
+              typename SAccElementFunction,
+              typename PComputeElementFunction,
+              typename OAccElementFunction>
+    CK_TILE_DEVICE auto operator()(const QDramBlockWindowTmp& q_dram_block_window_tmp, // M0*K0 tile
```
Could update some comments like this to reflect the new tile shapes in unified attention.

samremes commented on example/ck_tile/01_unified_attention/codegen/ops/__init__.py:
Also under the example, could move everything under the existing 01_fmha example, but have it in separate files if needed, naming it unified_attention, and avoid duplicating all the code.

samremes reviewed (commented):
I didn't go too much into the details of the attention implementation yet, but these are mostly comments on structure and style. There is also recently [merged PR](https://github.com/ROCm/composable_kernel/pull/3153) for fmha_v3 that may cause some refactoring, when merging the changes. Avoiding code duplication would make it much easier to see the differences to the v3 kernel and pipeline.

Chi-Chu319 committed:
remove if statements

Chi-Chu319 committed:
refactor

Chi-Chu319 committed:
Gemm dispatch

Chi-Chu319 committed:
dispatcher

Chi-Chu319 committed:
Fixes

Chi-Chu319 committed:
Merge branch 'develop' into tianxing/unified-attention

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+# fused multi-head attention
+
+This folder contains example for unified attention (fused multi-head attention) using ck_tile tile-programming implementation. It is a good example to demonstrate the usage of tile-programming API, as well as illustrate the new approach to construct a kernel template and instantiate it(them) while keeping compile time fast.
```
```suggestion
This folder contains examples for unified attention (fused multi-head attention) using the ck_tile tile-programming implementation. The examples demonstrate the usage of the tile-programming API, as well as the new approach to constructing kernel templates and instantiating them. 
```

Wasn't sure if the compile times were fast because of the new way that kernels were instantiated, or if it demonstrated a new way to keep compile times fast independent of instantiation.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+This folder contains example for unified attention (fused multi-head attention) using ck_tile tile-programming implementation. It is a good example to demonstrate the usage of tile-programming API, as well as illustrate the new approach to construct a kernel template and instantiate it(them) while keeping compile time fast.
+
+## build
```
Add an empty line below this

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+# you can replace <arch> with the appropriate architecture (for example gfx90a or gfx942) or leave it blank
+../script/cmake-ck-dev.sh  ../ <arch>
+make tile_example_unified_attention -j
+```
````
Add an empty line below this

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+## kernel
+The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
+
+There are 2 template parameters for this kernel template.
```
Add an empty line below

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+```
+This will result in an executable `build/bin/tile_example_unified_attention`
+
+## kernel
````
Put an empty line beneath all headers

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+This will result in an executable `build/bin/tile_example_unified_attention`
+
+## kernel
+The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
```
This is really confusing.

Do you mean the original version of CK or some pre-release version of CK Tile?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
+
+There are 2 template parameters for this kernel template.
+* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of differen ...
```
This is a really confusing paragraph. 

Is `FmhaPipeline` an enum of `block_tile_pipeline` class? If so, then it should be referred to as such. If not, can you explain what it is because it's not clear.

The rest reads like an internal note. I don't understand what it's trying to say. Can you clarify?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+There are 2 template parameters for this kernel template.
+* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of differen ...
+* `EpiloguePipeline` will modify and store out the result in the last phase. People usually will do lot of post-fusion at this stage, so we also abstract this concept. Currently we didn't do much thing at the epilogue stage but leave the room for future possible support.
```
```suggestion
* `EpiloguePipeline` is the last stage of the pipeline. It modifies and stores the result. Post-fusion can be done at this stage though the example only returns the result.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of differen ...
+* `EpiloguePipeline` will modify and store out the result in the last phase. People usually will do lot of post-fusion at this stage, so we also abstract this concept. Currently we didn't do much thing at the epilogue stage but leave the room for future possible support.
+
+## codegen
```
new line

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+* `EpiloguePipeline` will modify and store out the result in the last phase. People usually will do lot of post-fusion at this stage, so we also abstract this concept. Currently we didn't do much thing at the epilogue stage but leave the room for future possible support.
+
+## codegen
+To speed up compile time, we instantiate the kernels into separate file. In this way we can benefit from parallel building from CMake/Make system. This is achieved by `generate.py` script. Besides, you can look into this script to learn how to instantiate a kernel instance step by step, which is described in `FMHA_FWD_KERNEL_BODY` variable.
```
```suggestion
To speed up compile time, the `generate.py` script splits the kernel into separate files. This allows the example to take advantage of CMake/Make parallel building.  
```

What does `FMHA_FWD_KERNEL_BODY` describe exactly?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+## codegen
+To speed up compile time, we instantiate the kernels into separate file. In this way we can benefit from parallel building from CMake/Make system. This is achieved by `generate.py` script. Besides, you can look into this script to learn how to instantiate a kernel instance step by step, which is described in `FMHA_FWD_KERNEL_BODY` variable.
+
+## executable
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+To speed up compile time, we instantiate the kernels into separate file. In this way we can benefit from parallel building from CMake/Make system. This is achieved by `generate.py` script. Besides, you can look into this script to learn how to instantiate a kernel instance step by step, which is described in `FMHA_FWD_KERNEL_BODY` variable.
+
+## executable
+`tile_example_unified_attention` is the example executable, implemented in `fmha_fwd.cpp`. You can type `./bin/tile_example_unified_attention -?` to list all the arguments. Below is an example of the output (may subject to change)
```
```suggestion
Building `fmha_fwd.cpp` yields the `tile_example_unified_attention` executable. Use `./bin/tile_example_unified_attention -?` to list all the arguments. 
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+
+## executable
+`tile_example_unified_attention` is the example executable, implemented in `fmha_fwd.cpp`. You can type `./bin/tile_example_unified_attention -?` to list all the arguments. Below is an example of the output (may subject to change)
+```
````
newline above

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+`tile_example_unified_attention` is the example executable, implemented in `fmha_fwd.cpp`. You can type `./bin/tile_example_unified_attention -?` to list all the arguments. Below is an example of the output (may subject to change)
+```
+args:
+          -v    weather do CPU validation or not (default:1)
````
```suggestion
          -v    Perform CPU validation (default:1)
```

What are the possible values? Is the default to perform validation or not perform validation?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+          -v    weather do CPU validation or not (default:1)
+       -mode    kernel mode. 0:batch, 1:group (default:0)
+          -b    batch size (default:2)
+          -h    num of head, for q (default:8)
```
Is this number of heads? is there a lower or upper limit?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+          -b    batch size (default:2)
+          -h    num of head, for q (default:8)
+        -h_k    num of head, for k/v, -1 means equal to h (default:-1)
+                if not equal to h, then this is GQA/MQA case
```
```suggestion
        -h_k    Number of heads for k/v. When set to -1, the number of heads will be set to the same value as was set for q. (default:-1)
```

What is the GQA/MQA case?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                if not equal to h, then this is GQA/MQA case
+          -s    seqlen_q. if group-mode, means the average value of seqlen_q (default:3328)
+                total_seqlen_q = seqlen_q * batch, and seqlen_q per batch may vary
+                also with "-s=s0,s1,s2..." comma seperated int to set per batch seqlen(group-mode)
```
```suggestion
  seqlen_q. When in group-mode, this is the average value of seqlen_q. In group mode it can also be set to a series of comma-separated integers to set seqlen per batch.  (default:3328)```

Was the equation supposed to be total? The description mentions an average not a total:

total_seqlen_q = seqlen_q * batch 

Also, what can this be set to when not in group mode? Or is it not set at all in kernel mode?

Given that this can also be set to a series of comma-separated ints, why is the default such a large number (3328). Does the number have meaning?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                total_seqlen_q = seqlen_q * batch, and seqlen_q per batch may vary
+                also with "-s=s0,s1,s2..." comma seperated int to set per batch seqlen(group-mode)
+        -s_k    seqlen_k (including new key/value), -1 means equal to s (default:-1)
+                also with "-s_k=s0,s1,s2..." comma-separated ints to set seqlen per batch (group mode)
```
```suggestion
-s_k    seqlen_k. Set this to a list of comma-separated integers for seqlen per bathch, or set to -1 to assign the same values as seqlen_q (default:-1)
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+        -s_k    seqlen_k (including new key/value), -1 means equal to s (default:-1)
+                also with "-s_k=s0,s1,s2..." comma-separated ints to set seqlen per batch (group mode)
+     -s_qpad    seqlen_q stride between 2 batches (group-mode optional) (default:-1)
+                Provide positive strides per-batch to simulate physical padding on Q
```
```suggestion
seqlen_q stride size between 2 batches. Simulates physical padding on Q. Default -1.
```

What does a negative value, in particular the default value, do?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                also with "-s_k=s0,s1,s2..." comma-separated ints to set seqlen per batch (group mode)
+     -s_qpad    seqlen_q stride between 2 batches (group-mode optional) (default:-1)
+                Provide positive strides per-batch to simulate physical padding on Q
+     -s_kpad    seqlen_k stride between 2 batches, currently used in group-mode only  (default:-1)
```
need to say what -1 does

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                for kv-cache case, each batch [1,s,h,d]/[1,h,s,d] can have a stride
+                along seqlen, instead of packed, same as xformer kv_padding,
+                must be greater than or equal to s_k
+          -d    head dim for q, k (default:128)
```
```suggestion
          -d    head dimension for q, k (default:128)
```

Need to say if there's a lower or upper bound

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                along seqlen, instead of packed, same as xformer kv_padding,
+                must be greater than or equal to s_k
+          -d    head dim for q, k (default:128)
+        -d_v    head dim for v, -1 means equal to d (default:-1)
```
need to say whether there's a lower or upper bound

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+    -range_k    per-tensor quantization range of k. used if squant=1. (default:16)
+    -range_v    per-tensor quantization range of v. used if squant=1. (default:16)
+    -range_p    per-tensor quantization range of p [e^(s-m)]. used if squant=1. (default:1)
+    -range_o    per-tensor quantization range of o (p*v). used if squant=1. (default:16)
```
in general, you should state if there are lower or upper bounds on the values

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+    -range_o    per-tensor quantization range of o (p*v). used if squant=1. (default:16)
+     -squant    if using static quantization fusion or not. auto: fp8 will default use squant, other will not (default:auto)
+                0: no static quant(not implemented) 1: apply scale_p and scale_o with respect to P and O.
+                calculate scale_s, scale_p, scale_o according to range_q, range_k, range_v, range_p, range_o
```
```suggestion
When set to 0, no static quantization fusion is applies. When set to 1, scale_p and scale_o are applied with respect to P and O. When set to auto, only fp8 will apply static quantization. (default:auto)
```

Not sure how the last line about calculating those values fit in.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                0: no static quant(not implemented) 1: apply scale_p and scale_o with respect to P and O.
+                calculate scale_s, scale_p, scale_o according to range_q, range_k, range_v, range_p, range_o
+      -iperm    permute input (default:1)
+                if true, will be b*h*s*d, else b*s*h*d
```
```suggestion
When set to 1, input will be permuted with b*h*s*d. When set to 0, will be b*s*h*d. Default 1.
```

I'm assuming 1==true and 0==false

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                calculate scale_s, scale_p, scale_o according to range_q, range_k, range_v, range_p, range_o
+      -iperm    permute input (default:1)
+                if true, will be b*h*s*d, else b*s*h*d
+      -operm    permute output (default:1)
```
need the values and the results

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+      -operm    permute output (default:1)
+       -bias    n or 0, no bias (default:n)
+                e(lementwise) or 1, elementwise bias with 1*1*s*s. e:1, 1*h*s*s. e:2, b*h*s*s
+                a(libi) or 2, alibi with 1*h. a:1, b*h
```
```suggestion
Set to  n or 0 for no bias. Set to e or 1 for elementwise bias with 1*1*s*s. Set to e:1 for elementwise bias with 1*h*s*s. Set to e:2 for elementwise bias with b*h*s*s. Set to a or 2 for alibi bias with 1*h. Set to a:1 for alibi bias with b*h. Default is 0, no bias.```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+       -bias    n or 0, no bias (default:n)
+                e(lementwise) or 1, elementwise bias with 1*1*s*s. e:1, 1*h*s*s. e:2, b*h*s*s
+                a(libi) or 2, alibi with 1*h. a:1, b*h
+       -prec    data type. fp16/bf16/fp8/bf8 (default:fp16)
```
what type of data type? for what? Needs clarification

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                'b:l,r', bottom-r sliding window attn(swa) with FA style left right size
+                'xt:window_size', xformer style masking from top-left, window_size negative is causal, positive is swa
+                'xb:window_size', xformer style masking from bottom-r, window_size negative is causal, positive is swa
+                'g:y,x', generic attention mask coordinate with y/x size (only debug purpose for now)
```
```suggestion
 Set to 0 for no mask. Set to 1 top-left mask. Set to 2 for bottom-right mask. Set to 't' for top-left causal mask. Set to 'b' for bottom-right causal mask. Set to 't:l,r' for top-left sliding window attn(swa) with FA style left right size mask. Set to 'b:l,r' for bottom-r sliding window attn(swa) with FA style left right size mask. Set to 'xt:window_size' for an xformer style mask from top-left where setting window_size to a negative value results in a causal mask and setting it to a positive value with result in an swa mask. Set to 'xb:window_size' for an xformer style mask from bottom-right where setting window_size to a negative value results in a causal mask and setting it to a positive value with result in an swa mask. Set to  'g:y,x' for generic attention mask coordinates with y/x size (only for debugging purposes). Default is 0, no mask.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                'xt:window_size', xformer style masking from top-left, window_size negative is causal, positive is swa
+                'xb:window_size', xformer style masking from bottom-r, window_size negative is causal, positive is swa
+                'g:y,x', generic attention mask coordinate with y/x size (only debug purpose for now)
+    -vlayout    r for row-major(seqlen*hdim), c for col-major(hdim*seqlen) (default:r)
```
```suggestion
    -vlayout    Set to r for row-major. Set to c for column-major. (default:r)
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+                'xb:window_size', xformer style masking from bottom-r, window_size negative is causal, positive is swa
+                'g:y,x', generic attention mask coordinate with y/x size (only debug purpose for now)
+    -vlayout    r for row-major(seqlen*hdim), c for col-major(hdim*seqlen) (default:r)
+        -lse    0 not store lse, 1 store lse (default:0)
```
```suggestion
        -lse    Set to 0 to not store lse. Set to 1 to store lse (default:0)
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+        -lse    0 not store lse, 1 store lse (default:0)
+      -kname    if set to 1 will print kernel name (default:0)
+       -init    init method. ui, uniform random int, ni, normalized random int (default:uf)
+                uf, uniform random float, nf, normalized random float, tf, trig float, uf:q, quantization
```
```suggestion
 init method. Set to ui for a uniform random int. Set to ni for a normalized random int. Set to uf for a uniform random float. Set to nf for a normalized random float. Set to  tf for a trig float. Set to uf:q for quantization. Default uf.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+      -kname    if set to 1 will print kernel name (default:0)
+       -init    init method. ui, uniform random int, ni, normalized random int (default:uf)
+                uf, uniform random float, nf, normalized random float, tf, trig float, uf:q, quantization
+       -seed    random seed used for initializing input tensors. 0 for non-deterministic seed (default:11939)
```
Is there a reason for that default?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+       -seed    random seed used for initializing input tensors. 0 for non-deterministic seed (default:11939)
+  -drop_seed    seed for random number generator (default:1)
+-drop_offset    offset for random number generator (default:0)
+ -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)
```
I don't quite understand what this option is. Can you clarify?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+  -drop_seed    seed for random number generator (default:1)
+-drop_offset    offset for random number generator (default:0)
+ -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)
+ -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)
```
What is 1? What does the default do?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+-drop_offset    offset for random number generator (default:0)
+ -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)
+ -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)
+     -warmup    number of iterations before benchmark the kernel (default:5)
```
```suggestion
     -warmup    number of iterations before benchmarking the kernel (default:5)
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+ -drop_prefs    seed and offset values are present on GPU; 0 - host, 1 - device/GPU (default:0)
+ -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)
+     -warmup    number of iterations before benchmark the kernel (default:5)
+     -repeat    number of iterations to benchmark the kernel (default:20)
```
```suggestion
     -repeat    number of times to iterate when benchmarking the kernel (default:20)
```

For this and warmup, any upper or lower bounds should be mentioned.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+ -num_splits    number of splits for key/value. 0 to determine actual number by heuristic (default:1)
+     -warmup    number of iterations before benchmark the kernel (default:5)
+     -repeat    number of iterations to benchmark the kernel (default:20)
+       -json    0: No Json, 1: Dump Results in Json format (default:0)
```
```suggestion
       -json    Set to 1 to write out results in to a json file. (default:0, results are not written to a json file)
```

Where do the results go when set to 0

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+     -warmup    number of iterations before benchmark the kernel (default:5)
+     -repeat    number of iterations to benchmark the kernel (default:20)
+       -json    0: No Json, 1: Dump Results in Json format (default:0)
+   -jsonfile    json file name to dump results (default:fmha_fwd.json)
```
```suggestion
   -jsonfile    json file to which to write results when `-json` is set to 1 (default:fmha_fwd.json)
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+       -json    0: No Json, 1: Dump Results in Json format (default:0)
+   -jsonfile    json file name to dump results (default:fmha_fwd.json)
+ -q_eff_lens    Batch-mode only: per-batch effective seqlen for Q (exclude PAD) (default:"")
+                Comma-separated list of length 'b'. If empty, no override
```
I don't quite understand what this is. What is b in this case?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+ -q_eff_lens    Batch-mode only: per-batch effective seqlen for Q (exclude PAD) (default:"")
+                Comma-separated list of length 'b'. If empty, no override
+-kv_eff_lens    Batch-mode only: per-batch effective seqlen for KV (exclude PAD) (default:"")
+                Comma-separated list of length 'b'. If empty, no override
```
See previous comment.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+                Comma-separated list of length 'b'. If empty, no override
+-kv_eff_lens    Batch-mode only: per-batch effective seqlen for KV (exclude PAD) (default:"")
+                Comma-separated list of length 'b'. If empty, no override
+```
````
insert newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Example 2: `./bin/tile_example_unified_attention -b=1 -h=8 -s=16384 -d=64 -drop_prefs=1 -drop_seed=10 -drop_offset=1234` will run a fmha case with 
+  batch=1, nhead=8, sequence length=16384, hdim=64, drop_seed=0 (in GPU memory), drop_offset=1234 (in GPU memory) fp16 case
+
+## Padding Examples
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+Example 4 (Batch mode with effective lengths): `./bin/tile_example_unified_attention -mode=0 -b=2 -h=8 -s=2048 -s_k=2048 -d=128 -q_eff_lens=1024,1536 -kv_eff_lens=1024,1536` will run batch mode where all batches use 2048 as physical sequence length but have effective lengths of (1024, 1536) for Q and KV respectively.
+
+## support features
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Example 4 (Batch mode with effective lengths): `./bin/tile_example_unified_attention -mode=0 -b=2 -h=8 -s=2048 -s_k=2048 -d=128 -q_eff_lens=1024,1536 -kv_eff_lens=1024,1536` will run batch mode where all batches use 2048 as physical sequence length but have effective lengths of (1024, 1536) for Q and KV respectively.
+
+## support features
+Currently we are still in rapid development stage, so more features/optimizations will be coming soon.
```
```suggestion
This is still in the development phase and new features and optimizations are being added.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+## support features
+Currently we are still in rapid development stage, so more features/optimizations will be coming soon.
+
+### hdim
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Currently we are still in rapid development stage, so more features/optimizations will be coming soon.
+
+### hdim
+Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, within which `64`/`128` is better optimized. hdim should be multiple of 8, while seqlen_s can be arbitrary. For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)
```
```suggestion
Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, optimized for `64`/`128`. hdim should be set to a multiple of 8, while seqlen_s can be set to any value. 
```

Not sure what this means:

"For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)"

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### hdim
+Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, within which `64`/`128` is better optimized. hdim should be multiple of 8, while seqlen_s can be arbitrary. For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)
+
+### group/batch mode
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Currently we support `32/64/128/256` hdim for `fp16`/`bf16`, within which `64`/`128` is better optimized. hdim should be multiple of 8, while seqlen_s can be arbitrary. For hdim be arbitrary number, it can be support through padding kernel of `qr` pipeline (we didn't generate this in generate.py by default)
+
+### group/batch mode
+Currently we support both `batch mode` and `group mode` (or `varlen`, in FA's term), by setting `-mode` = `0` or `1`. In `group mode` different kind of attention mask is also supported(see below)
```
```suggestion
Currently we support both batch mode and group mode (or varlen, in FA's term). Different types of attention masking can be applied group mode.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### group/batch mode
+Currently we support both `batch mode` and `group mode` (or `varlen`, in FA's term), by setting `-mode` = `0` or `1`. In `group mode` different kind of attention mask is also supported(see below)
+
+### MQA/GQA
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Currently we support both `batch mode` and `group mode` (or `varlen`, in FA's term), by setting `-mode` = `0` or `1`. In `group mode` different kind of attention mask is also supported(see below)
+
+### MQA/GQA
+By setting `-h`(nhead for q) and `-h_k`(nhead for k/v) with different number, you can achieve MQA/GQA. Please pay attention that `h % h_K == 0` when you set different numbers.
```
This is very vague. Can it be clarified?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### MQA/GQA
+By setting `-h`(nhead for q) and `-h_k`(nhead for k/v) with different number, you can achieve MQA/GQA. Please pay attention that `h % h_K == 0` when you set different numbers.
+
+### input/output permute, and `b*s*3*h*d`
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+By setting `-h`(nhead for q) and `-h_k`(nhead for k/v) with different number, you can achieve MQA/GQA. Please pay attention that `h % h_K == 0` when you set different numbers.
+
+### input/output permute, and `b*s*3*h*d`
+If you look at the kernel argument inside `fmha_fwd_kernel.hpp`, we support providing arbitrary stride for seqlen(stride_q/k/v), nhead, batch of q/k/v matrix, hence it is very flexible to support `b*h*s*d` or `b*s*h*d` input/output permute. The `-iperm=0/1`, `-operm=0/1` is a convenient way to achieve this through the executable. We didn't provide a command-line arg to test `b*s*3*h*d` layout whi ...
```
```suggestion
The kernel argument inside `fmha_fwd_kernel.hpp` supports arbitrary strides for seqlen (stride_q/k/v), nhead, batch of q/k/v matrix, making it possible to support `b*h*s*d` or `b*s*h*d` input/output permutations.  The `-iperm=0/1`, `-operm=0/1` options is a convenient way to achieve this. There is no option to test `b*s*3*h*d` layout which is by default used by torch/FA, but it's possible to do this by setting the proper `stride_q/k/v` value as `3*h*d`.
```

Does this have to be set in the code itself? 

also: in general it's not a good idea to say something is trivial. Not all users might find something trivial.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### input/output permute, and `b*s*3*h*d`
+If you look at the kernel argument inside `fmha_fwd_kernel.hpp`, we support providing arbitrary stride for seqlen(stride_q/k/v), nhead, batch of q/k/v matrix, hence it is very flexible to support `b*h*s*d` or `b*s*h*d` input/output permute. The `-iperm=0/1`, `-operm=0/1` is a convenient way to achieve this through the executable. We didn't provide a command-line arg to test `b*s*3*h*d` layout whi ...
+
+### attention bias
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+If you look at the kernel argument inside `fmha_fwd_kernel.hpp`, we support providing arbitrary stride for seqlen(stride_q/k/v), nhead, batch of q/k/v matrix, hence it is very flexible to support `b*h*s*d` or `b*s*h*d` input/output permute. The `-iperm=0/1`, `-operm=0/1` is a convenient way to achieve this through the executable. We didn't provide a command-line arg to test `b*s*3*h*d` layout whi ...
+
+### attention bias
+Attention bias is supported with the layout of `1*1*s*s`(similiar to input/output, different layout can be supported by changing the stride value for bias, or even extend to `b*h*s*s`) and bias value in float number.
```
This is very confusing. Can this be clarified?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+alibi is supported
+
+### lse
+For training kernels, "log sum exp" need to store out in forward and used in backward. We support this by setting `-lse=1`
```
This is confusing. Can it be clarified a bit?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### attention bias
+Attention bias is supported with the layout of `1*1*s*s`(similiar to input/output, different layout can be supported by changing the stride value for bias, or even extend to `b*h*s*s`) and bias value in float number.
+
+### alibi
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### alibi
+alibi is supported
+
+### lse
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### lse
+For training kernels, "log sum exp" need to store out in forward and used in backward. We support this by setting `-lse=1`
+
+### vlayout
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+For training kernels, "log sum exp" need to store out in forward and used in backward. We support this by setting `-lse=1`
+
+### vlayout
+We support v matrix in both row-major(`seqlen*hdim`) and col-major(`hdim*seqlen`). Since the accumulate(reduce) dimension for V is along `seqlen`, for current AMD's mfma layout which expect each thread to have contiguous register holding pixels along reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major, there a ...
```
```suggestion
We support V matrix in both row-major and column-major. Since the accumulate (reduce) dimension for V is along `seqlen`, and AMD's mfma layout expects each thread to have contiguous registers holding pixels along the reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major. There are many factors that may affect the overall performance. We still provide the `-vlayout=r/c` here to switch/test between different layouts.
```

When you say pixels, do you mean data?

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### vlayout
+We support v matrix in both row-major(`seqlen*hdim`) and col-major(`hdim*seqlen`). Since the accumulate(reduce) dimension for V is along `seqlen`, for current AMD's mfma layout which expect each thread to have contiguous register holding pixels along reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major, there a ...
+
+### attention mask
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+We support v matrix in both row-major(`seqlen*hdim`) and col-major(`hdim*seqlen`). Since the accumulate(reduce) dimension for V is along `seqlen`, for current AMD's mfma layout which expect each thread to have contiguous register holding pixels along reduce dimension, it's easier to support col-major V layout. However, the performance of col-major is not necessarily faster than row-major, there a ...
+
+### attention mask
+we support `causal mask` and `sliding window attention(swa)` mask in both batch and group mode, either from top-left or bottom-right.
```
```suggestion
Causal and sliding window attention(swa) masks in both batch and group mode are supported, both from either top-left or bottom-right.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+### attention mask
+we support `causal mask` and `sliding window attention(swa)` mask in both batch and group mode, either from top-left or bottom-right.
+Underneath, we unify the mask expression into `generic attention mask coordinate`, providing an uniformed approach for each batch to locate the corresponding pixel need to be masked out.
```
```suggestion
The mask expression is unified into generic attention mask coordinates, providing an uniform approach for each batch to locate the corresponding pixel need to be masked out.
```

Is this pixels? Is this example for image masking? If that's the case, it should be made clear earlier on.

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+Underneath, we unify the mask expression into `generic attention mask coordinate`, providing an uniformed approach for each batch to locate the corresponding pixel need to be masked out.
+![](misc/gamc.png)
+
+Since FA/xformer style with window_size_left/right is more popular, we accept window_size as parameter and convert that internally to our generic coordinate(this coordinate can express more cases). Below shows some example of how to achieve different kind of mask through cmdline.
```
```suggestion
Since FA/xformer style with window_size_left/right is more popular, we accept window_size as parameter and convert that internally to our generic coordinate. The following are some examples of how to achieve different kind of mask through cmdline:
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+Note FA use bottom-right by default to express swa case, here we require you explicitly specify top-left/bottom-right.
+
+### dropout
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### dropout
+TBD
+
+### sequence padding and variable length support
```
newline

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+### sequence padding and variable length support
+We support sequence padding and variable-length processing in both batch and group modes fmha forward to handle real-world scenarios where sequences have different lengths.
+
+**Group Mode Padding**: Use `-s_qpad` and `-s_kpad` to specify physical stride between batches, enabling padded layouts. Each batch can have different logical sequence lengths (`-s`, `-s_k`) but use larger physical strides for memory alignment.
```
```suggestion
**Group Mode Padding**: Use `-s_qpad` and `-s_kpad` to specify physical strides between batches and enable padded layouts. Each batch can have different logical sequence lengths (`-s`, `-s_k`). Use larger physical strides for memory alignment.
```

spolifroni-amd commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+Both approaches optimize memory access patterns while supporting flexible sequence length requirements commonly found in transformer inference scenarios.
+
+## FP8 experimental support
```
newline

spolifroni-amd reviewed (changes requested):
The readme needs some work.

spolifroni-amd reviewed (changes requested):
The readme needs some work.

Chi-Chu319 committed:
Merge branch 'develop' into tianxing/unified-attention

Chi-Chu319 committed:
Update example/ck_tile/42_unified_attention/README.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Chi-Chu319 commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+This folder contains examples for unified attention (fused multi-head attention) using the ck_tile tile-programming implementation. The examples demonstrate the usage of the tile-programming API, as well as the new approach to constructing kernel templates and instantiating them. 
+
+## build
+```
````
```suggestion
## build

```
```

Chi-Chu319 commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+# you can replace <arch> with the appropriate architecture (for example gfx90a or gfx942) or leave it blank
+../script/cmake-ck-dev.sh  ../ <arch>
+make tile_example_unified_attention -j
+```
+This will result in an executable `build/bin/tile_example_unified_attention`
````
```suggestion
```

This will result in an executable `build/bin/tile_example_unified_attention`
```

Chi-Chu319 commented on example/ck_tile/42_unified_attention/README.md:
````diff
@@ -0,0 +1,159 @@
+make tile_example_unified_attention -j
+```
+This will result in an executable `build/bin/tile_example_unified_attention`
+
+## kernel
+The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
````
```suggestion
## kernel

The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
```

Chi-Chu319 commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,159 @@
+
+## kernel
+The kernel template is `unified_attention.hpp`, this is the grid-wise op in old ck_tile's terminology. We put it here purposely, to demonstrate one can construct a kernel by using various internal component from ck_tile. We may still have an implementation under ck_tile's include path (in the future) for the kernel template.
+
+There are 2 template parameters for this kernel template.
+* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of differen ...
```
```suggestion
There are 2 template parameters for this kernel template.

* `FmhaPipeline` is one of the block_tile_pipeline(under `include/ck_tile/tile_program/block_tile_pipeline`) which is a performance critical component. Indeed, we did a lot of optimization and trials to optimize the pipeline and may still workout more performance pipeline and update into that folder. People only need to replace this pipeline type and would be able to enjoy the benefit of different performant implementations (stay tuned for updated pipeline(s)).
```

Chi-Chu319 committed:
Update example/ck_tile/42_unified_attention/README.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Chi-Chu319 committed:
Update example/ck_tile/42_unified_attention/README.md

Chi-Chu319 committed:
Update example/ck_tile/42_unified_attention/README.md

Chi-Chu319 committed:
Update example/ck_tile/42_unified_attention/README.md

illsilin committed:
Merge branch 'develop' into tianxing/unified-attention

asleepzzz commented:
please remove fmha_bwd_known_fails_gfxxxx.txt since you don't need them

poyenc commented:
please resolve the merge conflicts

poyenc commented on example/ck_tile/42_unified_attention/unified_attention_impl.hpp:
```diff
@@ -0,0 +1,176 @@
+    index_t kBlockQ = Kernel::kBlockQ;
+    assert(args.num_queries_per_kv == Kernel::num_queries_per_kv &&
+           "argument num_queries_per_kv must equal compiled num_queries_per_kv");
+    assert(args.BLOCK_SIZE == Kernel::BLOCK_SIZE &&
```
The `unified_attention_args` does not have the `BLOCK_SIZE` attribute. Please review if this `assert()` is necessary.

poyenc commented on example/ck_tile/42_unified_attention/unified_attention_impl.hpp:
```diff
@@ -0,0 +1,176 @@
+           "argument num_queries_per_kv must equal compiled num_queries_per_kv");
+    assert(args.BLOCK_SIZE == Kernel::BLOCK_SIZE &&
+           "argument BLOCK_SIZE must equal compiled BLOCK_SIZE");
+    assert(kBlockQ == kBlockM / args.num_queries_per_kv &&
```
Add missing quailfer `Kernel::` for accessing those attributes: `Kernel::kBlockM`

poyenc commented on example/ck_tile/42_unified_attention/example_unified_attention.cpp:
```diff
@@ -0,0 +1,680 @@
+    args.num_head_q         = problem.nhead_q;
+    args.num_queries_per_kv = num_queries_per_kv;
+    args.page_blk_size      = problem.page_blk_size;
+    args.mask_type          = 2;
```
The `mask_type` is hardcoded and ignores the command-line option, so we cannot tell whether the unmasked kernels work correctly. Please update this.

poyenc commented on example/ck_tile/42_unified_attention/README.md:
```diff
@@ -0,0 +1,161 @@
+     -repeat    number of iterations to benchmark the kernel (default:20)
+       -json    0: No Json, 1: Dump Results in Json format (default:0)
+   -jsonfile    json file name to dump results (default:fmha_fwd.json)
+ -q_eff_lens    Batch-mode only: per-batch effective seqlen for Q (exclude PAD) (default:"")
```
Please remove `q_eff_lens`/`kv_eff_lens` from the README.md if they are not intended to be supported.

ammallya commented:
Error importing due to merge conflicts – please reopen the PR on ROCm/rocm-libraries

illsilin commented:
all development must be done in the monorepo now. please create a new pr there: https://github.com/ROCm/rocm-libraries

aghamari committed:
Fixes compilation issues for unified attention

aghamari committed:
Add unified attention d64/GQA-8 kernel instances and fix BLOCK_SIZE for small head dims
The unified attention kernel previously only supported head_size=128
with MHA (NumQPerKV=1). This change adds support for head_size=64 with
GQA-8 (NumQPerKV=8), which is the configuration used by models like
DeepSeek-V3/R1 (64 query heads, 8 KV heads, head_dim=64).
Changes:
- Add 4 new kernel instance files for d64/GQA-8:
  unified_attention_d64_{bf16,fp16}_{nmask,mask}_gqa8.cpp
- Add d64/GQA-8 dispatch path in unified_attention.cpp
- Fix BLOCK_SIZE (kPageBlockSize) in unified_attention_kernel_traits:
  compute from HEAD_SIZE instead of hardcoding 32. For HeadSize<=64,
  BLOCK_SIZE must be 64 to guarantee NumIssues>=1 on gfx950. With
  128-bit vector loads (KVector=8), LaneGroups*NumWarps=128 exceeds
  kPageBlockSize=32 when HeadSize=64, causing a division-by-zero in
  the LDS tile descriptor constexpr evaluation.
